### PR TITLE
[TRAFODION-2117] add SQL support of EXCEPT

### PR DESCRIPTION
**Pull request by traflm**
initial support of SQL syntax EXCEPT

### CODE REVIEW

**Comment by traflm** [comment]
jenkins, retest

File: core/sql/optimizer/BindRelExpr.cpp [hunk 1]
```diff
@@ -3035,6 +3035,98 @@ RelExpr *Intersect::bindNode(BindWA *bindWA)
...
+    *CmpCommon::diags() << DgSqlCode(-4014);
+    bindWA->setErrStatus();
+    return this;
+  }
+
+  // Join the columns of both sides. 
+  //
+  if(CmpCommon::getDefault(MODE_SPECIAL_4) != DF_ON)
+  {
+    *CmpCommon::diags() << DgSqlCode(-3022)    // ## INTERSECT not yet supported
```
Comment (DaveBirdsall) [inline core/sql/optimizer/BindRelExpr.cpp]: Comment should say EXCEPT (you can fix it on a later check-in)

File: core/sql/optimizer/BindRelExpr.cpp [hunk 2]
```diff
@@ -3035,6 +3035,98 @@ RelExpr *Intersect::bindNode(BindWA *bindWA)
...
+
+  if (child(0)->getGroupAttr()->isStream() &&
+      child(1)->getGroupAttr()->isStream()){
+    *CmpCommon::diags() << DgSqlCode(-4159)
+                << DgString0(fmtdList1) << DgString1(fmtdList2);
+    bindWA->setErrStatus();
+    return this;
+  }
+
+  // Needs to be removed when supporting get_next for INTERSECT
```
Comment (DaveBirdsall) [inline core/sql/optimizer/BindRelExpr.cpp]: Ditto

**Comment by DaveBirdsall** [comment]
Nicely done!

**Comment by traflm** [comment]
Jenkins, extra tests

**Comment by traflm** [comment]
fail at TEST042, something with SQL query cache, which should not be introduced by this PR. I will make the change of @DaveBirdsall 's comments in next chance of commit code.

File: core/sql/optimizer/BindRelExpr.cpp [hunk 3]
```diff
@@ -3035,6 +3035,98 @@ RelExpr *Intersect::bindNode(BindWA *bindWA)
 // LCOV_EXCL_STOP
 
 // -----------------------------------------------------------------------
+// member functions for class Except 
+// -----------------------------------------------------------------------
+
+// LCOV_EXCL_START - cnu
```
Comment (sureshsubbiah) [inline core/sql/optimizer/BindRelExpr.cpp]: This coverage line should be removed in a subsequent checkin.

File: core/sql/optimizer/RelExpr.cpp [hunk 4]
```diff
@@ -6630,6 +6630,23 @@ const NAString Intersect::getText() const
 }
 
 // -----------------------------------------------------------------------
+// // member functions for class Except
+// // -----------------------------------------------------------------------
+Except::Except(RelExpr *leftChild,
+             RelExpr *rightChild)
+: RelExpr(REL_EXCEPT, leftChild, rightChild)
+{ setNonCacheable(); }
```
Comment (sureshsubbiah) [inline core/sql/optimizer/RelExpr.cpp]: Why is this noncacheable?
Also please check if we can do showshape on a statement with EXCEPT

File: core/sql/optimizer/BindRelExpr.cpp [hunk 5]
```diff
@@ -3035,6 +3035,98 @@ RelExpr *Intersect::bindNode(BindWA *bindWA)
...
+  if (leftTable.getDegree() != rightTable.getDegree()) {
+    // 4014 The operands of an intersect must be of equal degree.
+    *CmpCommon::diags() << DgSqlCode(-4014);
+    bindWA->setErrStatus();
+    return this;
+  }
+
+  // Join the columns of both sides. 
+  //
+  if(CmpCommon::getDefault(MODE_SPECIAL_4) != DF_ON)
```
Comment (sureshsubbiah) [inline core/sql/optimizer/BindRelExpr.cpp]: I forget now, why is this guarded by MODE_SPECIAL_4 ?

**Comment by sureshsubbiah** [comment]
Looks good. If you prefer we can address the cache and shape comments later (if there is indeed a change needed). Will wait to hear from @traflm before merging.

File: core/sql/optimizer/BindRelExpr.cpp (hunk 5 above)
Comment (traflm) [inline core/sql/optimizer/BindRelExpr.cpp]: this is not required, but new features not well QAed, I think it is easier to be adopted if protected by CQD :-)
After QA tested, we can externalize all these new features: INTERSECT and EXCEPT

File: core/sql/optimizer/RelExpr.cpp (hunk 4 above)
Comment (traflm) [inline core/sql/optimizer/RelExpr.cpp]: I don't know either, but SET operations like UNION is also set as non cacheable. So I feel we should keep it now. I will try to study more on this later.

showshape works:
>>showshape select eno,ename,dno from t021 except select * from t022;
I
control query shape hybrid_hash_join(scan(path 'TRAFODION.SEABASE.T021',
forward, blocks_per_access 1 , mdam off),
scan(path 'TRAFODION.SEABASE.T022', forward, blocks_per_access 1
, mdam off));

--- SQL operation complete.
>>

**Comment by traflm** [comment]
Thanks @DaveBirdsall @sureshsubbiah for carefully review and the findings. 
I wish to change them in my next check in and get this merged first, since they are typos in comments. For other comments, please see my reply above.

**Pull request merged**